crypto/merklesignature/merkleSignatureScheme.go

```diff
@@ -0,0 +1,278 @@
+	SignerContext struct {
+		_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+		FirstValid uint64           `codec:"rnd"`
+		Interval   uint64           `codec:"iv"`
+		Tree       merklearray.Tree `codec:"tree"`
+	}
+
+	// Verifier is used to verify a merklesignature.Signature produced by merklesignature.Secrets.
+	Verifier [MerkleSignatureSchemeRootSize]byte
+
+	//KeyRoundPair represents an ephemeral signing key with it's corresponding round
+	KeyRoundPair struct {
+		_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+		Round uint64               `codec:"rnd"`
+		Key   *crypto.FalconSigner `codec:"key"`
+	}
+)
+
+// CryptoPrimitivesID is an identification that the Merkle Signature Scheme uses a subset sum hash function
+// and a falcon signature scheme.
+var CryptoPrimitivesID = uint16(0)
+
+// Errors for the merkle signature scheme
+var (
+	ErrStartBiggerThanEndRound           = errors.New("cannot create Merkle Signature Scheme because end round is smaller then start round")
+	ErrDivisorIsZero                     = errors.New("received zero Interval")
+	ErrNoStateProofKeyForRound           = errors.New("no stateproof key exists for this round")
+	ErrSignatureSchemeVerificationFailed = "merkle signature verification failed"
+)
+
+// New creates secrets needed for the merkle signature scheme.
+// This function generates one key for each round within the participation period [firstValid, lastValid] (inclusive bounds)
+// which holds round % interval == 0.
+// In case firstValid equals zero then signer will generate all keys from (0,Z], i.e will not generate key for round zero.
+func New(firstValid, lastValid, interval uint64) (*Secrets, error) {
+	if firstValid > lastValid {
+		return nil, ErrStartBiggerThanEndRound
+	}
+	if interval == 0 {
+		return nil, ErrDivisorIsZero
+	}
+
+	if firstValid == 0 {
+		firstValid = 1
+	}
+
+	// calculates the number of indices from first valid round and up to lastValid.
+	// writing this explicit calculation to avoid overflow.
+	numberOfKeys := lastValid/interval - ((firstValid - 1) / interval)
+
+	keys, err := KeysBuilder(numberOfKeys)
+	if err != nil {
+		return nil, err
+	}
+	tree, err := merklearray.BuildVectorCommitmentTree(&CommittablePublicKeyArray{keys, firstValid, interval}, crypto.HashFactory{HashType: MerkleSignatureSchemeHashFunction})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Secrets{
+		ephemeralKeys: keys,
+		SignerContext: SignerContext{
+			FirstValid: firstValid,
+			Interval:   interval,
+			Tree:       *tree,
+		},
+	}, nil
+}
+
+// GetVerifier can be used to store the commitment and verifier for this signer.
+func (s *Secrets) GetVerifier() *Verifier {
+	return s.SignerContext.GetVerifier()
+}
+
+// GetVerifier can be used to store the commitment and verifier for this signer.
+func (s *SignerContext) GetVerifier() *Verifier {
+	ver := [MerkleSignatureSchemeRootSize]byte{}
+	ss := s.Tree.Root().ToSlice()
+	copy(ver[:], ss)
```

Comment:
Should we log some error and return a pointer to an empty verifier in that case?

Reply:
I think that a better solution would be to keep the code as is, but adding a verification in `merklesignature.New` that validates that the configured tree produces a hash of the desired length.
Then, add a comment in `GetVerifier()` that saying that you're using that assumption here as well.
( this might be ok, since the tree object is internal, and the logic of the code prevents the hash length from changing )

Reply:
The `Root()` method returns a `GenericDigest` which is already defined to be of size `SumhashDigestSize`, same as `MerkleSignatureSchemeRootSize`. Is that not enough to feel confident in this copy?
I'm not sure what check we could add, since the way nodes (and the root) are stored in the tree is within slices of `SumhashDigestSize` anyway.